lib/posix-netlink: Introduce netlink library

andreittr opened the pull request:
### Description of changes

This introduces `posix-netlink` a socket library handling the `AF_NETLINK` protocol, and an intermediary for other libraries to register handlers for specific netlink families.

Based on earlier work by @skuenzer and @jschlumpp .

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

skuenzer committed:
lib/syscall_shim: Pretty print decoding of netlink

This commit introduces initial print format and decoding definitions for
netlink sockets.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-netlink: Introduce library skeleton

This change contains the empty library and Kconfig options.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-netlink: Import ABI headers from Linux

This change imports <linux/netlink.h> providing netlink constants and
data structs. File is imported with trivial minor edits to make it work
with Unikraft's build system.

Source of import:
Repository: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git
Tag: v5.10 (2c85ebc57b)
Path: include/uapi/linux/netlink.h

Checkpatch-Ignore: COMMIT_LOG_LONG_LINE
Checkpatch-Ignore: COMMIT_LOG_USE_LINK
Checkpatch-Ignore: BLOCK_COMMENT_STYLE
Checkpatch-Ignore: POINTER_LOCATION
Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: LONG_LINE
Checkpatch-Ignore: LONG_LINE_COMMENT
Checkpatch-Ignore: SPACE_BEFORE_TAB
Checkpatch-Ignore: BIT_MACRO
Checkpatch-Ignore: SPACING
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-netlink: Initial implementation

Initial implementation of netlink library.

Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: LONG_LINE
Co-authored-by: Andrei Tatar <ttr@unikraft.io>
Signed-off-by: Andrei Tatar <ttr@unikraft.io>
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-netlink: Return ENOTSUP in ioctl & poll

This change has ioctl return ENOTSUP and poll warn stubbed.

Checkpatch-Ignore: SPACING
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

jschlumpp committed:
lib/posix-netlink: Initialize msg_name structure

This part of the structure is used by glibc to determine whether the
response came from the kernel.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
lib/posix-netlink: Only write addr if ptr provided

Applications can set the address pointer to NULL if they are not interested
in the source.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

andreittr committed:
lib/posix-netlink: Adapt to new posix-socket API

This change adapts netlink to the new polling API.

Checkpatch-Ignore: SPACING
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

michpappas commented on lib/posix-netlink/include/uk/netlink/driver.h:
```diff
@@ -0,0 +1,133 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
According to our conventions this should be `include/uk/netlink.h`. Since it's only intended for drivers though, it could probably be in `include/uk/netlink/driver.h`.

andreittr replied:
The latter sounds good :+1:

michpappas commented on lib/posix-netlink/include/linux/netlink.h:
```diff
@@ -0,0 +1,354 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
```
Since this is an imported header, ideally there should be some traceability info on the commit message.

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+	/* FIXME: PID is here like port number and must be unique. It just
+	 *        starts with PID if free.
+	 */
+	nl_ctx->nl_pid = getpid();
```
Let's use `uk_sys_getpid()` to avoid making syscalls from the kernel's context

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+{
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	struct uk_streambuf *nlbuf;
+
```
Perhaps it would be worth having a `UK_ASSERT(s)` before deriving the netlink context. Ditto for pointer parameters in the rest of ops.

andreittr replied:
Both these asserts are done implicitly by [`posix_sock_get_node()`](https://github.com/unikraft/unikraft/blob/staging/lib/posix-socket/include/uk/socket_driver.h#L99) which `sock2nlctx` uses.
And in general (at least across the file/sock/fs stack), the 1st object argument to operations (akin to `this` or `self`) is always guaranteed non-NULL by caller inlines. The op function may at most wish to assert it's of the correct type.

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+
+	if (*addr_len < sizeof(*nl_addr)) {
+		/* addr_len is too small, only inform about needed size */
+		goto out;
```
If I'm reading [getsockname(2)](https://man7.org/linux/man-pages/man2/getsockname.2.html) correctly, this case should also return `EINVAL`. Also this should probably be `unlikely()`. Same for a few more input checks in other ops.

mogasergiu replied:
The comment I think is misleading. I see the man says:
>  The returned address is truncated if the buffer provided is too small; in this case, addrlen will return a value. greater than was supplied to the call.

My understanding is that it does write partial data.

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+
+	ret = nl_handle(nl_ctx, buf, len);
+	if (unlikely(ret < 0))
+		return -EINVAL;
```
Perhaps this should be returning the error value of `nl_handle()`.

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+	nl_addr = (struct sockaddr_nl *)dest_addr;
+	if (unlikely(nl_addr->nl_family != AF_NETLINK || nl_addr->nl_pid != 0))
+		return -EHOSTUNREACH;
+
```
Also a check that the socket is bound would be worth here:
```c
if (unlikely(!nl_ctx->bound))
    return -EDESTADDRREQ;
```

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+	     nlh = NLMSG_NEXT(nlh, len)) {
+		uk_pr_debug("Call handler for msg %p (driver %s)\n",
+			    nlh, nl_ctx->drv->libname);
+		ret = nl_ctx->drv->ops->handle(nl_ctx, nlh);
```
Without checking the value of `ret`, this continues execution and returns the error code of the last message sent.

michpappas commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,339 @@
+	if (msg->msg_name && msg->msg_namelen == sizeof(struct sockaddr_nl)) {
+		nl_addr = msg->msg_name;
+		nl_addr->nl_family = AF_NETLINK;
+		nl_addr->nl_pid = 0;
```
I take it that for now we only support process to kernel communication hence pid is always set to 0?

michpappas reviewed (changes requested):
Hi @andreittr, minor comments only.

mogasergiu commented on lib/posix-netlink/Config.uk:
```diff
@@ -0,0 +1,8 @@
+	bool "posix-netlink: Abstraction for netlink protocols"
+	select LIBPOSIX_SOCKET
+	help
+	  This microlibrary allows other microlibraries to register a
```
So, I've always thought that we use LK style for KConfig and I've even pointed it out to you in the past :sweat: . But recently it's been brought up to me that that is actually not the case and we actually do use tab instead of 2 whitespaces for the help message indentation.
CC: @michpappas

mogasergiu commented on lib/posix-netlink/include/uk/netlink_protocol.h:
```diff
@@ -0,0 +1,133 @@
+		__section("." _POSIX_NETLINK_PROTOCOL_SECNAME(lib, prot))\
+	static struct posix_netlink_protocol				\
+	_POSIX_NETLINK_PROTOCOL_DRVRNAME(lib, prot) = {			\
+		.protocol = prot,					\
```
Should `prot`/`vops` here be surrounded by `()`?

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+	/* AF_NETLINK sockets must be created with SOCK_RAW or SOCK_DGRAM */
+	if (unlikely(((type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)) != SOCK_RAW) &&
+		     ((type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)) != SOCK_DGRAM))) {
+		err = EINVAL;
```
Why do we use positive errno values in this function only to then return the negated version anyway?

andreittr replied:
Might have to do with what error values the `create` netlink op is meant to return; the docs in the driver header are misleading and need updating. Will convert to negative and enforce netlink drivers return negative error codes.

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+	UK_ASSERT(nl_ctx);
+	UK_ASSERT(nl_ctx->drv);
+	UK_ASSERT(nl_ctx->drv->ops);
+	UK_ASSERT(nl_ctx->drv->ops->handle);
```
so why is only `handle` asserted for but `close`/`create` are actively checked for? maybe this should also be documented that only `handle` is mandatory?

mogasergiu commented on lib/posix-netlink/include/uk/netlink_protocol.h:
```diff
@@ -0,0 +1,133 @@
+
+#define nlbuf_reserve(nlbuf, len)	uk_streambuf_reserve((nlbuf), (len))
+#define nlbuf_align(nlbuf, align)					       \
+	uk_streambuf_reserve((nlbuf), ALIGN_UP(uk_streambuf_seek((nlbuf)),     \
```
Why does this "call" (I know it's a macro) `uk_streambuf_seek` twice in a row?

mogasergiu commented on lib/posix-netlink/include/uk/netlink_protocol.h:
```diff
@@ -0,0 +1,133 @@
+
+static inline void nlbuf_send(struct nl_ctx *ctx, struct uk_streambuf *nlbuf)
+{
+	return uk_mbox_post(ctx->nl_recvqueue, nlbuf);
```
This confused me and actually made me check if `uk_mbox_post` actually does return a value, but it doesn't. Should we remove the `return`?

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+	const struct sockaddr_nl *nl_addr;
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+
+	if (unlikely(nl_ctx->bound))
```
I see the manual says that `bind` returns `EINVAL` in this case?
>  EINVAL The socket is already bound to an address.

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+
+	/* We only support the client to connect to us: the kernel */
+	if (nl_addr->nl_pid != 0)
+		return -EHOSTUNREACH;
```
`bind` does not return `EHOSTUNREACH` (although I have seen syscalls that in practice do return undocumented error codes... looking at you `EFAULT` from `copy_from/to_user`).
Should it instead be `EADDRNOTAVAIL`? But it also sounds a bit like a more unikraft specific case.

andreittr replied:
It is indeed a unikraft-specific case for which we need to pick an error ourselves, as the upstream API leaves it undefined; since `EADDRNOTAVAIL` is at least _documented_ in `bind(2)` it's more likely to be gracefully handled by workloads and thus probably the better choice :+1: .

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+		return -EINVAL;
+
+	nl_addr = (struct sockaddr_nl *)dest_addr;
+	if (unlikely(nl_addr->nl_family != AF_NETLINK || nl_addr->nl_pid != 0))
```
`sendto` does not return `EHOSTUNREACH` according to manual... but it does according to POSIX :exploding_head: 
But I think the error here should be:
```C
if (unlikely(nl_addr->nl_family != AF_NETLINK))
	return -EAFNOSUPPORT;
if (unlikely(nl_addr->nl_pid != 0))
	return -EINVAL;  // or ECONNREFUSED
```

andreittr replied:
I think this is also one case where unikraft is special; in my (possibly naive) understanding of netlink as a _best effort_ protocol, sending a message to a destination that's not listening should silently drop the packet.
Thus in theory we should _not_ error out here, rather drop the request immediately (since in the current impl there cannot be anyone but the kernel listening). However, I'd rather have it throw an error to notify us of this missing feature, than have it stay silent, however correct the latter may be.
Again as for `bind`, error is up to us: `EINVAL` is already too overloaded, `ECONNREFUSED` is not mentioned in the `sendto(2)` manpage (nor does it technically apply here, since netlink sockets are not connection-oriented), so `EHOSTUNREACH` seems reasonable to me (esp since POSIX _does_ document it for `sendto`).

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+		memcpy(iovec[i].iov_base, src, cpylen);
+
+		left -= cpylen;
+		src  += cpylen;
```
void pointer arithmetic :thinking:

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+	if (unlikely(!nlbuf))
+		return -EIO;
+
+	cpylen = copy2iovec(msg->msg_iov, msg->msg_iovlen,
```
Should we validate the iovec?
```C
	if (unlikely(msg->msg_iovlen > 0 && !msg->msg_iov))
		return -EFAULT;
```

But also, something I am realising just now about both `uk_mbox_recv` and `uk_mbox_post` that are used across this driver.
They are blocking! Shouldn't the socket driver always be non-blocking? i.e. use the `try` variants.
(also `uk_mbox_recv` has its own return value and `EAGAIN` would then probably be more suitable)

andreittr replied:
ukfile drivers (including sockets) are allowed to block if they can't help it, and the rest of the file stack works OK. Netlink sockets in particular, because their handlers atm run synchronously with the request, are even more amenable to this, since there's never any "waiting" for replies -- requests generate and write the replies before completion.
If the implementation evolves to support more features, and nonblocking operation will be genuinely required, we can revisit this. In the meantime I'll leave an explanatory comment.

Edit: on 2nd thought I should at least use `_try` versions on `recv`, to prevent silly workloads from deadlocking themselves if they decide to read an empty netlink socket. Polling/notification is still out of scope and only intended for full async support later if needed.

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,309 @@
+			    nlbuf_data(nlbuf), nlbuf_len(nlbuf));
+	nlbuf_free(nlbuf);
+	uk_pr_debug("Message received\n");
+	return (ssize_t)cpylen;
```
Are we sure the message does not get truncated? In that case, according to manual for `recvmsg`, we should set `MSG_TRUNC` flag

mogasergiu commented on lib/posix-netlink/protocol_list.ld:
```diff
@@ -0,0 +1,10 @@
+		PROVIDE(posix_netlink_protocol_list_end = .);
+	}
+}
+INSERT BEFORE .data;
```
Should these entries be writable? If not, then:
```
INSERT AFTER .rodata;
```

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -254,6 +255,16 @@ static ssize_t nl_recvmsg(posix_sock *s, struct msghdr *msg,
 	uk_pr_debug("Message received\n");
+
+	/* Fill name structure */
+	if (msg->msg_name && msg->msg_namelen == sizeof(struct sockaddr_nl)) {
```
If buffer is larger, is it not still valid?
If yes then maybe we should use
```C
if (msg->msg_name && msg->msg_namelen >= sizeof(struct sockaddr_nl))
```
and in that case, I think we should update `msg_namelen` to how much we actually wrote to the bigger buffer.

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -277,6 +277,9 @@ static ssize_t nl_recvfrom(posix_sock *s,
 	struct nl_ctx *nl_ctx = sock2nlctx(s);
 	size_t cpylen;
 
+	if (from && !fromlen)
```
This commit made me check and actually should we also check for `!buf` and return `-EFAULT`?

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -295,6 +295,18 @@ static int nl_sockopt(const posix_sock *s __unused,
+		    int request __unused,
+		    void *argp __unused)
+{
+	return -ENOTSUP;
```
Why do we need a separate commit for this and shouldn't it actually be `ENOTTY`?

andreittr replied:
> Why do we need a separate commit for this

To preserve history and attribution.

> shouldn't it actually be ENOTTY?

I think that error only applies to ioctls specific to character device files; it probably should be `-EINVAL` though, as that's the generic "I don't support the args you passed".

nderjung commented on lib/posix-netlink/Config.uk:
```diff
@@ -0,0 +1,10 @@
+	help
+	  This microlibrary allows other microlibraries to register a
+	  netlink protocol family, such as NETLINK_ROUTE, for their netlink
+	  module implementation. This library regsiters the POSIX socket API
```
registers*

andreittr committed:
lib/ukfile: Add missing assert header include

This change adds an include for <uk/assert.h> to <uk/file/iovutil.h>, as
the latter explicitly calls asserts.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-netlink: Refactor & API completeness

This change is a minor refactor of the netlink socket implementation,
completing its implementation of the posix-socket API, with stubs where
needed. In addition, the netlink driver API is better documented, and
several previous issues and cornercases fixed.

Checkpatch-Ignore: MACRO_ARG_REUSE
Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed update:
- rebasing existing commits on top of current staging -- no changes
- 1 trivial fix commit w/ a missing include
- 1 minor refactor commit addressing review comments and completing the socket API

mogasergiu commented on lib/posix-netlink/netlink.c:
```diff
@@ -0,0 +1,443 @@
+	if (unlikely(!addr || addr_len < sizeof(*nl_addr)))
+		return -EINVAL;
+
+	nl_addr = (struct sockaddr_nl *)addr;
```
Should we use `DECONST`?

andreittr replied:
This is a true cast, not just getting rid of `const`: `const struct sockaddr *` -> `struct sockaddr_nl *`.
But now that you mention it, `nl_addr` should be const too, will fix.

andreittr commented:
Pushed addressing comments

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (approved):
LGTM
Reviewed-by: Michalis Pappas <michalis@unikraft.io>
Approved-by: Michalis Pappas <michalis@unikraft.io>
